Update metrics.md; add telemetry events

irrationalagent opened the pull request:
Resolves #58 
Resolves #180 

Here's an updated metrics md for Sandy (and Devin). I'll wait till https://github.com/mozilla-lockbox/lockbox-ios/pull/296 is merged, then rebase and add the events in this PR

irrationalagent commented:
@sandysage if you have comments on the doc let me know and I'll make the changes

irrationalagent commented:
@sashei was there any particular way that you were able to verify that the (e.g.) `CopyAction` telemetry events were being recorded directly? I tried grepping for it in the simulator data after performing the action but that didn't seem to work

sashei commented:
@irrationalagent hrm ok, I will investigate.

sashei commented:
@irrationalagent it looks like we don't collect events for simulator builds; on line 56 in `TelemetryAction.swift`, just switch `isCollectionEnabled` from false to true.

irrationalagent commented:
looked at that line several times and was like 'nah, that's not it' 🤦‍♂️

sashei commented:
@irrationalagent ok, I got to fool around with this a little bit before james got here this morning, so I may be wrong about it, but I have a very strong suspicion ;) let me know how you go & I can take a closer look if that's not it.

irrationalagent commented:
Note: in order to debug these events I changed these lines in TelemetryAction.swift, but reverted the changes before committing here:

```swift
#if DEBUG
        telemetryConfig.isCollectionEnabled = true // this one
        telemetryConfig.isUploadEnabled = false
        telemetryConfig.updateChannel = "debug"
```
```swift
    lazy var telemetryActionListener: AnyObserver<TelemetryAction> = {
        return Binder(self) { target, action in
            target.telemetry.recordEvent(
                    category: TelemetryEventCategory.action.rawValue,
                    method: action.eventMethod.rawValue,
                    object: action.eventObject.rawValue,
                    value: action.value,
                    extras: action.extras
            )
            print (    // I added this print statement
              TelemetryEventCategory.action.rawValue,
              action.eventMethod.rawValue,
              action.eventObject.rawValue,
              action.value,
              action.extras
            );
        }.asObserver()
    }()
}
```
Note that 
(1) I'm not sure if the first change was actually required and 
(2) even though adding the print statement logged the correct event data to xcode console and at the right times, I'm not sure it actually means that the events are being recorded. It could just mean they're getting to the spot at which they are supposed to be recorded (at least IIUC).

irrationalagent commented:
@sashei no rush on this review. @devinreams if want to take it that's fine too

devinreams commented:
>no rush on this review. @devinreams if want to take it that's fine too

Thanks, I'll test it out locally and make sure I can reproduce your console logging, at the least.

I'll defer to the experts on the code review itself. In the meantime, a fun photo that I'll think of every time I see this PR number:

![311](https://user-images.githubusercontent.com/49511/39496193-a31f5c04-4d5a-11e8-8a38-41e94c54e70d.jpg)

irrationalagent commented:
a couple of other notes:
1. The `revealPassword` event always fires when someone views an item detail, but the `value` field logs whether it was revealed (couldnt figure out how to log it in the reveal-only case)

2. I'm using the `value` field now to log settings values etc, let me know if we need to some kind of validation there

3. I had to change the test that references `faqButton` since that event object doesn't exist anymore

sashei commented on lockbox-ios/Action/FxAAction.swift:
```diff
@@ -38,6 +63,39 @@ enum FxAError: Error {
     case Unknown
 }
 
+extension FxAError: TelemetryAction {
```
The FxA error never gets sent as an `Action`, so this extension won't ever get invoked. I think we may need to circle back with @jhugman about how to do this logging in our current "old"-FxA world.

irrationalagent commented:
went ahead and removed the fxa bits for now. looks like we *may* need some more tests here too ?

sashei commented:
@irrationalagent yesss, I'd say so. Every `[something]Action.swift` has a corresponding `[something]ActionSpec.swift` file that you can extend with the `TelemetryAction` bits.

devinreams commented:
From today's standup, @sashei and @sandysage are re-requested for final review.

Nice work @irrationalagent 👏

sashei commented on lockbox-ios/Action/RouteAction.swift:
```diff
@@ -29,6 +29,34 @@ enum MainRouteAction: RouteAction {
+        return nil
+    }
+
+    var extras: [String: Any?]? {
```
These extras are not covered by any tests.

irrationalagent replied:
@sasha how do i do something like: 

```swift
it("returns item id as extra") {
     expect(MainRouteAction.detail(itemId: "sfddfs").extras).to(equal([ExtraKey.itemid.rawValue: "sfddfs"]))
}

sashei replied:
@irrationalagent maybe try .to(contain(x)) ?

irrationalagent replied:
😞 feel defeated by this... I've tried a many things but can't get something to work... especially these seem like they should work but im obviously in the dark when it comes to typing in swift

```swift
expect(MainRouteAction.detail(itemId: "sfddfs").extras).to(equal(["itemid": "sfddfs"]))
```
`Cannot convert value of type 'Predicate<[String : String]>' to expected argument type 'Predicate<[String : Any?]>'`

```swift
expect(MainRouteAction.detail(itemId: "sfddfs").extras!.Values).to(contain("sfddfs"))
```
or
```swift
expect(MainRouteAction.detail(itemId: "sfddfs").extras!.Values).to(equal(["sfddfs"]))
```
`Static member 'Values' cannot be used on instance of type '[String : Any?]'`

I considered converting the schema(?) for extras to be 

```swift 
var extras: [String: String?]? { get }
```

but I'd prefer not to since some extras might be integers in the future and it would involve changing a bunch of other lines as well... plz halp

sashei commented on lockbox-ios/Action/RouteAction.swift:
```diff
@@ -38,6 +66,38 @@ enum SettingRouteAction: RouteAction {
+        return nil
+    }
+
+    var extras: [String: Any?]? {
```
not covered by tests

sashei commented:
@irrationalagent maybe try `.to(contain(x))` ?

sashei commented:
heya @irrationalagent , let me know if this update makes sense to you! swifty silliness :(

irrationalagent commented:
Ah, thanks @sashei TIL. couldn't quite get there myself, thanks for bringing this PR over the line.
